enhancements/installer/single-node-installation-bootstrap-in-place.md

```diff
@@ -0,0 +1,669 @@
+
+For bootstrap-in-place model to work in cloud environemnt we need to mitigate the following gaps:
+1. The bootstrap-in-place model relay on the live ISO environment as a place to write bootstrapping files so that they don't end up on the real node.
+Optional mitigation: We can mimic this environment by mounting some directories as tmpfs during the bootstrap phase.
+2. The bootstrap-in-place model uses coreos-installer to write the final Ignition to disk along with the RHCOS image.
+Optional mitigation: We can boot the machine with the right RHCOS image for the release.
+Instead of writing the Ignition to disk we will use the cloud credentials to update the node Ignition config in the cloud provider.
+
+### Check control plane replica count in `create ignition-configs`
+
+Instead of adding a new installer command, we could use the current command for generating Ignition configs
+`create ignition-configs` to generate the `bootstrap-in-place-for-live-iso.ign` file,
+by adding logic to the installer that check the number of replicas for the control
+ plane (in the `install-config.yaml`) is `1`.
+This approach might conflict with CRC/SNC which also run openshift-install with a 1-replica control plane.
+
+### Use `create ignition-configs` with environment variable to generate the `bootstrap-in-place-for-live-iso.ign`
+
+We also considered adding a new environment variable `OPENSHIFT_INSTALL_EXPERIMENTAL_BOOTSTRAP_IN_PLACE`
+for marking the new path under the `ignition-configs` target.
+We decided to add `single-node-ignition-config` target to in order to gain:
+1. Allow us to easily add different set of validations (e.g. ensure that the number of replicas for the control plane is 1).
+2. We can avoid creating unnecessary assets (master.ign and worker.ign).
+3. Less prune to user errors than environment variable.
+
+### Alternatives for managing the operating system installation step
+
+Before settling on using a manual process, we identified several
+options for installing the operating system during deployment.
+
+1. We could give the user instructions to ssh to the host and run a
+   command.
+
+   This is the option we selected. It requires additional work for
+   users running the installer by hand, and will require some changes
+   to the CI jobs for single-node deployments, but is the most
+   flexible and does not commit us to maintaining and API we have
+   rushed to design.
+
+2. We could add an unsupported experimental environment variable to
+   all installer builds, so that the user could pass information
+   through the OpenShift installer to the CoreOS installer.
+
+   The risk of using an environment variable is that it would become a
+   de facto API that we must support over a long period of time. This
+   is especially likely since we are collaborating closely with users
+   and cannot guarantee that they will not build automation of their
+   own around the installer.
```

Comment:
This "it will be a defacto API with backwards-compat requirements" caveat is true of the `ssh`-command approach too, isn't it?  Can we make whatever is selected dev/tech-preview for the first round to make it more likely that folks understand that there may be rough edges and that we might end up needing compat-breaking workflow changes?

Reply:
> This "it will be a defacto API with backwards-compat requirements" caveat is true of the `ssh`-command approach too, isn't it? Can we make whatever is selected dev/tech-preview for the first round to make it more likely that folks understand that there may be rough edges and that we might end up needing compat-breaking workflow changes?

+1 - well pointed out

Reply:
Resolve by the update which adds an API